File: scoreboard/scoreboard.py
#!/usr/bin/env python3
# sourcery off

import sys
import collections
import functools

# log = open(sys.argv[1]).read().split("\n")
# lines = [[int(v) for v in s.split()] for s in log]
# teams, logs = lines[1]
# cases = lines[2:]

Sub = collections.namedtuple("Sub", ("time", "team", "problem", "input", "score"))
SubId = collections.namedtuple("SubId", ("team", "problem", "input"))
Team = collections.namedtuple("Team", ("id", "score", "penalty"))
Case = collections.namedtuple("Case", ("n_teams", "subs"))

line_vals = [[int(v) for v in line.split()] for line in open(sys.argv[1], "r")]
line_vals = [l for l in line_vals if l]  # rm empty
n_cases = line_vals.pop(0)[0]
cases = []
for ci in range(n_cases):
    vals = line_vals.pop(0)
    n_teams, n_logs = vals
    subs = []
    for li in range(n_logs):
        log = line_vals.pop(0)
        subs.append(Sub(*log))
    cases.append(Case(n_teams, subs))


def cmp_team(a, b):
    # a <  b  -1
    # a == b   0
    # a >  b   1
    if a.score < b.score:
        return -1
    elif a.score > b.score:
        return 1

    # penalty
    if a.penalty > b.penalty:
        return -1
    elif b.penalty > a.penalty:
        return 1

    # num order
    if a.id > b.id:
        return -1
    else:
        return 1


dbg = False
for ci, case in enumerate(cases):
    case_n = ci + 1

    seen_subs = set()
    teams = []
    for tid in range(1, case.n_teams + 1):
        if dbg: print("team", tid)
        score = 0
        penalty = 0
        for sub in case.subs:
            if sub.team != tid:
                continue

            if dbg: print(" - score", score)
            key = SubId(sub.team, sub.problem, sub.input)
            if key not in seen_subs and sub.score:
                score += sub.input * 100
                seen_subs.add(key)
                penalty += sub.time
                if dbg: print(" - penalty", sub.time)

        teams.append(Team(tid, score, penalty))

    if dbg: print(f"{case_n} scores:", " ".join(f"{t.id}={t.score}" for t in teams))
    t_ranked = sorted(teams, key=functools.cmp_to_key(cmp_team), reverse=True)
    print(f"Case #{case_n}:", " ".join(str(t.id) for t in t_ranked))  # team id

File: scoreboard/go/scoreboard.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"sort"
	"strconv"
	"strings"
)

func check(err error) {
	if err != nil {
		panic(err)
	}
}

type SubId struct {
	team int
	problem int
	input int
}

type Sub struct {
	time int
	team int
	problem int
	input int
	score int

	id SubId
}

type Team struct {
	id int
	score int
	penalty int
}

type Case struct {
	nTeams int
	subs []Sub
}

func cmpTeam(a, b Team) bool {
	if a.score < b.score { return true}
	if a.score > b.score { return false }

	// penalty
	if a.penalty > b.penalty { return true }
	if b.penalty > a.penalty { return false }

	// num
	if a.id > b.id {
		return true
	} else {
		return false
	}
}

func main() {
	data, err := ioutil.ReadFile(os.Args[1])
	check(err)

	sData := string(data)
	lines := strings.Split(sData, "\n")
	lineVals := make([][]int, 0, len(lines))
	for _, line := range lines {
		if len(line) == 0 {
			continue
		}

		vals := make([]int, 0, 5)
		sVals := strings.Split(line, " ")
		for _, val := range sVals {
			iVal, err := strconv.Atoi(val)
			check(err)
			vals = append(vals, iVal)
		}
		lineVals = append(lineVals, vals)
	}

	nCases := lineVals[0][0]
	cases := make([]Case, 0, nCases)
	pos := 1
	for ci := 0; ci < nCases; ci++ {
		vals := lineVals[pos]
		pos += 1
		nTeams := vals[0]
		nLogs := vals[1]
		subs := make([]Sub, 0, nLogs)
		for li := 0; li < nLogs; li++ {
			log := lineVals[pos]
			pos += 1
			sub := Sub{
				log[0], log[1], log[2], log[3], log[4],
				SubId{log[1], log[2], log[3]},
			}
			subs = append(subs, sub)
		}
		cases = append(cases, Case{nTeams, subs})
	}

	for ci, cse := range cases {
		caseN := ci + 1

		seenSubs := make(map[SubId]bool, len(cse.subs))
		teams := make([]Team, 0, cse.nTeams)
		for tid := 1; tid <= cse.nTeams; tid++ {
			score := 0
			penalty := 0
			for _, sub := range cse.subs {
				if sub.team != tid {
					continue
				}

				if _, ok := seenSubs[sub.id]; !ok && sub.score == 1 {
					score += sub.input * 100
					seenSubs[sub.id] = true
					penalty += sub.time
				}
			}

			teams = append(teams, Team{tid, score, penalty})
		}

		sort.SliceStable(teams, func(i, j int) bool {
			return !cmpTeam(teams[i], teams[j])
		})

		tidStrings := make([]string, 0, cse.nTeams)
		for _, team := range teams {
			tidStrings = append(tidStrings, strconv.Itoa(team.id))
		}

		fmt.Printf("Case #%d: %s\n", caseN, strings.Join(tidStrings, " "))
	}
}
